octodns/provider/yaml.py

```diff
@@ -28,8 +28,52 @@ class YamlProvider(BaseProvider):
         # (optional, default True)
         enforce_order: true
         # Whether duplicate records should replace rather than error
-        # (optiona, default False)
+        # (optional, default False)
         populate_should_replace: false
+        # The filename used to load split style zones, False means disabled.
+        # When enabled the provider will search for zone records split across
+        # multiple YAML files in the directory with split_extension appended to
+        # the zone name. split_extension should include the `.`
+        # See "Split Details" below for more information
+        # (optional, default False, . is the recommended best practice when
+        # enabling)
+        split_extension: false
+        # Disable loading of the primary zone .yaml file. If split_extension
+        # is defined both split files and the primary zone .yaml will be loaded
+        # by default. Setting this to true will disable that and rely soley on
+        # split files.
+        # (optional, default False)
+        split_only: false
+        # When writing YAML records out to disk with split_extension enabled
+        # each record is written out into its own file with .yaml appended to
+        # the name of the record. This would result in files like `.yaml` for
+        # the apex and `*.yaml` for a wildcard. If your OS doesn't allow such
+        # filenames or you would prefer to avoid them you can enable
+        # split_catchall to instead write those records into a file named
+        # `$[zone.name].yaml`
+        # (optional, default False)
+        split_catchall: false
+
+    Split Details
+    -------------
+
+    All files are stored in a subdirectory matching the name of the zone
+    (including the trailing .) of the directory config. It is a recommended
+    best practice that the files be named RECORD.yaml, but all files are
+    sourced and processed ignoring the filenames so it is up to you how to
+    organize them.
+
+    With `split_extension: .` the directory structure for the zone github.com.
+    managed under directory "zones/" would look like:
+
+    zones/
+      github.com./
+        .yaml
```

Comment:
See the `split_catchall` option. It's defaulting to `False` which means each record name gets its own file with the data for that name. OSX and Linux have no problems with files named `.yaml` and `*.yaml` and it seems cleaner to do it this way. If you use the SplitYamlProvider shim it'll get the previous/expected behavior.

Not sure about Windows, but either way the catchall behavior can be enabled by setting `split_catchall` and that file will be created with the APEX and/or wildcard data.

Should be documented in the param doc up above. I started to mention it down here as well, but the doc was getting pretty long.

Reply:
Ah gotcha. Should `split_catchall` default to `true` then to match the previous `SplitYamlProvider` behavior? I don't have a preference either way personally.

Reply:
The `[zone-name][ext]/$[zone-name].yaml` kind of confused me the first time I ran across it during this refactor, I haven't used split provider before, so that lead me to default to `false`. I assumed it had something to do with filename support on OSes so figured it's best to let people who are worried about that enable it provided they weren't coming from SplitYamlProvider where it was the default behavior.

I don't feel strongly about it, since I don't use it or plan to. So if I'm outvoted that's fine with me. Assuming you prefer true, @yzguy can be the tiebreaker :grin:

Reply:
So from my quick poking around (as I've never used the SplitYamlProvider either before today)

The user specified extension (`split_extension` in this case) is definitely related to #657, I assume that also applied to directories.

The `$` thing is actually pretty annoying regardless of OS because you need to escape it or the OS will think it's a variable, regardless it seems the previous behavior is it's only created whenever the record is `''` or `*`, it's not a user configurable thing.

So in that case, I'd think to preserve the current behavior of the SplitYamlProvider, whenever a user enables splitting (with `split_extension`), it should create the same catchall file if the record is `''` or `*`, if a user does not want it, they can disable it (which would be one of the new features of this change) with `split_catchall: false`

I would have to vote for `split_catchall: true` in that case, final answer :P

Reply:
> It would almost make more sense to just have the "catchall" be [zone-name][ext]/[zone-name].yaml, and the '' and * go in there, and we don't use any characters that the OS does special things with (variables, hidden files, etc.) in the file name

Mostly agree. Not really sure where the $ thing came from, I guess it's not a legal "name" and therefore could never exist from a record `zone.name.zone.name.` is a legal fqdn and would result in `[zone-directory]/[zone.name].yaml`. Thus you'd have the name based thing in that file along with the catchalls. Why someone would want it to I have no clue, but 🤷 

> These are likely too big of behavior changes but I'm thinking the ideal set up is

My main goal is to keep `SplitYamlProvder` unchanged and `YamlProvider` 100% backwards compatible. I'm pretty flexible as far as what the "new" behaviors are.

> disable_zonefile would not be needed, whenever split_zone is enabled, the zone directory is used

My thinking was to allow the option to partially split things, some split files and some in the main zone file. 

At GitHub I likely would have done that for the large dynamic records and then put all the simple/secondary stuff into the main file. The other use-case, which I may have done there too, would be to put important stuff in its own files with CODEOWNERS requiring specific reviews. This could even be extended to have per team files. 

Granted it could also be accomplished with just split files and using whatever filenames you want. No reason both split and the primary are required. The names of the files are actually ignored. Split is really just "directory"

I'll think about removing `disable_zonefile` in favor of split_extension turning off the zone file. That seems to make sense.

I will also enable catchall by default, but I think i'll leave it with the `$` to avoid having the odd/edge case of having to merge `zone.name.zone.name.` into that file if it happened to exist.

Reply:
So the final decision is to go with `$zone.yaml` format? If so, should the example be updated per my suggestion at the top?

Reply:
Good catch. Forgot that bit when I flipped the default. Fix pushed.

Reply:
The example also needs to show `$github.yaml` rather than just `.yaml` since we're defaulting to that behavior with `split_catchall: true`?